[DOCS] Refactored index-templates topic. (#59737)
elastic/elasticsearch: docs/reference/indices/put-component-template.asciidoc

[[indices-component-template]]
=== Put component template API
++++
<titleabbrev>Put component template</titleabbrev>
++++

Creates or updates a component template. 
Component templates are building blocks for constructing <<indices-templates,index templates>>  

> that specify index <<mapping,mappings>>, <<index-modules-settings,settings>>, 

Reviewer: I'd consider splitting this into two sentences.

and <<indices-aliases,aliases>>. 

[source,console]
--------------------------------------------------
PUT _component_template/template_1
{
  "template": {
    "settings": {
      "number_of_shards": 1
    },
    "mappings": {
      "_source": {
        "enabled": false
      },
      "properties": {
        "host_name": {
          "type": "keyword"
        },
        "created_at": {
          "type": "date",
          "format": "EEE MMM dd HH:mm:ss Z yyyy"
        }
      }
    }
  }
}
--------------------------------------------------
// TESTSETUP

//////////////////////////

[source,console]
--------------------------------------------------
DELETE _component_template/template_*
--------------------------------------------------
// TEARDOWN

//////////////////////////

[[put-component-template-api-request]]
==== {api-request-title}

`PUT /_component_template/<component-template>`

[[put-component-template-api-desc]]
==== {api-description-title}

An index template can be composed of multiple component templates. 

> To use a component template, you specify it the index template's `composed_of` list.

Reviewer: nit
```suggestion
To use a component template, specify it the index template's `composed_of` list.
```

You cannot use component templates independently, 

> they are only applied to new data streams and indices if they are part of a matching index template. 

Reviewer: This is a comma splice.

> You cannot use component templates independently

`Independently` seems a little unclear here. In some ways, modularity = independence.

It may be more clear to say you cannot apply or use component templates with an index template.

`version`::
(Optional, integer)
Version number used to manage component templates externally.
This number is not automatically generated or incremented by {es}.

`_meta`::
(Optional, object)
Optional user metadata about the component template. May have any contents.
This map is not automatically generated by {es}.

[[put-component-template-api-example]]
==== {api-examples-title}

===== Component template with index aliases

You can include <<indices-aliases,index aliases>> in a component template.

[source,console]
--------------------------------------------------
PUT _component_template/template_1
{
  "template": {
    "settings" : {
        "number_of_shards" : 1
    },
    "aliases" : {
        "alias1" : {},
        "alias2" : {
            "filter" : {
                "term" : {"user" : "kimchy" }
            },
            "routing" : "kimchy"
        },
        "{index}-alias" : {} <1>
    }
  }
}
--------------------------------------------------
<1> the `{index}` placeholder in the alias name will be replaced with the
actual index name that the template gets applied to, during index creation.

[[applying-component-templates]]
===== Applying component templates

> Component templates on their own don't apply or interact with data streams or indices at all.

Reviewer: ```suggestion
You cannot directly apply a component template to a data stream or index.
```

> To take effect, they must be used as part of an index template's `composed_of` field. See
> the <<indices-templates,index templates>>.

Reviewer: ```suggestion
To be applied, a component template must be included in an index template's `composed_of` list. See <<indices-templates>>.
```

[[component-templates-version]]
===== Component template versioning

You can use the `version` parameter to add an optional version number to a component template. External
systems can use these version numbers to simplify template management.

> The `version` parameter is completely optional and not automatically generated by {es}.

Reviewer: May be helpful to state that it's not _used_ by ES either.

To unset a `version`, replace the template without specifying one.

[source,console]
--------------------------------------------------
PUT /_component_template/template_1
{
  "template": {
    "settings" : {
        "number_of_shards" : 1
    }
  },
  "version": 123
}
--------------------------------------------------

> To check the `version`, you can use the <<getting-component-templates,get component template>> API.

Reviewer: nit
```suggestion
To check the `version`, you can use the <<getting-component-templates,get component template API>>.
```

[[component-templates-metadata]]
===== Component template metadata

You can use the `_meta` parameter to add optional metadata to a component template. This is a

> user-defined map that can contain any data. This data will be stored in the cluster state however,

Reviewer: ```suggestion
user-defined object that can contain any data. This data will be stored in the cluster state however,
```

so keeping it short is preferrable.

> The `_meta` parameter is completely optional and not automatically generated by {es}.

Reviewer: Again, may be helpful to state it's not used by ES either.
